### PR TITLE
feat: autocompletion for the `preset` config

### PR DESCRIPTION
**Pull request by DamianGlowala**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

PR provides autocompletion for the available preset names, and yet allowing to enter arbitrary values for the custom ones.

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/types/nitro.ts [hunk 1]
```diff
@@ -113,6 +113,20 @@ export interface NitroOptions extends PresetOptions {
...
+    | 'cloudflare'
+    | 'digital-ocean'
+    | 'firebase'
+    | 'heroku'
+    | 'layer0'
+    | 'netlify'
+    | 'render-com'
+    | 'stormkit'
+    | 'vercel'
+    | (string & {})
```
Comment (danielroe) [inline src/types/nitro.ts:129]: can we infer from the actual list of presets - e.g. `keyof typeof presets`?

Comment (DamianGlowala) [reply src/types/nitro.ts:129]: Sure! How would you approach the difference between camel case and dash case though?

image

Comment (danielroe) [reply src/types/nitro.ts:129]: How about excluding presets with capitals like this?

```ts
type DeCamel<T extends string> = T extends Lowercase<T> ? T : never
```

We could also write a slightly more complex type to _convert_ camel -> kebab, but I think that's probably overkill.

Comment (DamianGlowala) [reply src/types/nitro.ts:129]: That works great, but does it mean we would only autocomplete preset names without capitals?

Comment (danielroe) [reply src/types/nitro.ts:129]: Threw this together for fun. And definitely not to use here, but ...

```ts
type UppercaseLetter = 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' | 'L' | 'M' | 'N' | 'O' | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'W' | 'X' | 'Y' | 'Z'
type KebabCase<T extends string> = T extends Lowercase<T>
  ? T
  : T extends `${infer Start}${UppercaseLetter}${string}`
    ? Start extends Lowercase<Start>
      ? T extends `${Start}${infer U}${infer Rest}`
        ? Start extends ''
          ? `${Lowercase<U>}${KebabCase<Rest>}`
          : `${Start}-${Lowercase<U>}${KebabCase<Rest>}`
        : never
      : never
    : never
```

Comment (danielroe) [reply src/types/nitro.ts:129]: Oh, good point... might be worth just hard-coding 😆

Comment (DamianGlowala) [reply src/types/nitro.ts:129]: That was quick! :)

Found a shorter one which also seems to work fine, any thoughts?
https://stackoverflow.com/questions/66140451/typescript-add-kebab-case-types-form-actual-camel-case-keys
```
type Kebab<T extends string, A extends string = ""> =
    T extends `${infer F}${infer R}` ?
    Kebab<R, `${A}${F extends Lowercase<F> ? "" : "-"}${Lowercase<F>}`> :
    A
```
image

Comment (danielroe) [reply src/types/nitro.ts:129]: Oh, nice! Seems a much more reasonable level of complexity.

**Review by pi0 — APPROVED** [review]
LOVE IT!

**Pull request merged**